python/pyarrow/pandas_compat.py

```diff
@@ -294,9 +288,29 @@ def _column_name_to_strings(name):
     return str(name)
 
 
+def _index_level_name(index, i, column_names):
+    """Return the name of an index level or a default name if `index.name` is
+    None or is already a column name.
+
+    Parameters
+    ----------
+    index : pandas.Index
+    i : int
+
+    Returns
+    -------
+    name : str
+    """
+    if index.name is not None and index.name not in column_names:
```

Comment:
Should we be concerned about the linear search for `index.name not in column_names`? If so, let's create a set outside the loop below that we can check so that we don't need to do a full scan of the column names for every index column.

Reply:
I did some timings, and conversion to a set typically takes twice the time of a single search in the list. So you already need to have 3 index levels to benefit from this, and I don't think this is the typical use case? 
So I would personally leave it as is, but can certainly also easily add the suggestion.

Reply:
Fine by me.